httptransport/client/differ.go

```diff
@@ -122,19 +145,35 @@ func (c *HTTP) LatestUpdateOperations(ctx context.Context) (map[string]uuid.UUID
 	}
 	switch res.StatusCode {
 	case http.StatusOK:
+		m := make(map[string][]driver.UpdateOperation)
+		if err := json.NewDecoder(res.Body).Decode(&m); err != nil {
+			return nil, err
+		}
+		// check for etag, if exists store the value and add returned map
+		// to cache
 		if v := res.Header.Get("etag"); v != "" && !strings.HasPrefix(v, "W/") {
-			c.diffValidator.Store(v)
+			cache.Lock()
+			cache.uo = m
+			cache.validator = v
+			cache.Unlock()
 		}
+		return m, nil
 	case http.StatusNotModified:
-		return nil, ErrUnchanged
+		// received conditional response, serve map from cache
+		cache.RLock()
+		// make a copy
+		m := map[string][]driver.UpdateOperation{}
+		for u, ops := range cache.uo {
+			o := make([]driver.UpdateOperation, len(ops), len(ops))
+			copy(o, ops)
+			log.Printf("%v %v", o, ops)
+			m[u] = o
+		}
+		cache.RUnlock()
+		return m, nil
 	default:
-		return nil, fmt.Errorf("%v: unexpected status: %s", u.Path, res.Status)
-	}
-	m := make(map[string]uuid.UUID)
-	if err := json.NewDecoder(res.Body).Decode(&m); err != nil {
-		return nil, err
 	}
-	return m, nil
+	return nil, fmt.Errorf("%v: unexpected status: %s", req.URL.Path, res.Status)
```

Comment:
In the above switch, the first case returns the map directly, and the second case makes a copy. This might be better expressed by adding a `Copy` method to the cache object, having the cases acquire and defer the lock, move this error return into the default case, and have this line return `cache.Copy(), nil`

Reply:
good call, missed that we store the returned value. Copy method sounds like a good idea.

Reply:
I added a `cache.Set` to remove any lock usage from the caller, as it got kinda weird that the caller needed to use the lock only sometimes.